README.md
[![CircleCI](https://circleci.com/gh/ouattararomuald/numbering-plan/tree/master.svg?style=shield)](https://circleci.com/gh/ouattararomuald/numbering-plan/tree/master)
[![](https://img.shields.io/badge/code--style-square-green.svg)](https://github.com/square/java-code-styles)
[![](https://img.shields.io/maven-central/v/com.ouattararomuald/numbering.svg)](https://search.maven.org/search?q=g:com.ouattararomuald%20a:numbering)
[![](https://img.shields.io/nexus/s/https/oss.sonatype.org/com.ouattararomuald/numbering.svg)](https://oss.sonatype.org/content/repositories/snapshots/)

# numbering-plan

This library helps you migrate exiting phone numbers format to new phone number formats.

## Usage

First you start by defining the migration rules for you country

```kotlin
val ivoryCoastPlanFactory: CountryPlan = CountryPlan.Builder()
    .setOldPhoneNumberSize(8)
    .setInternationalCallingCode("225")
    .setDigitMapperPosition(Position.START) // Position.END
    .setMigrationType(MigrationType.PREFIX) // MigrationType.POSTFIX
    .setPrefixesMapper(
      mapOf(
        "07" to "07", // Orange
        "08" to "07", // eg: 08 XX XX XX => 07 08 XX XX XX (if MigrationType.prefix is used) => 08 XX XX XX 07 (if MigrationType.postfix is used)
        "09" to "07",
        "04" to "05", // MTN
        "05" to "05",
        "06" to "05",
        "01" to "01", // MOOV
        "02" to "01",
        "03" to "01"
      )
    ).build()
```

Then you pass those rules to `NumberingPlan` and you can start the migration:

```kotlin
val ivoryCoastPlanFactory: CountryPlan = ...
val numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
val newPhoneNumbers = numberingPlan.migrate(mapOf(
  "userId-1" to listOf("08060709"),
  "userId-2" to listOf("06060709"),
  "userId-6" to listOf("zezae/03-060-701", "01020304"),
  "userId-3" to listOf("03060701"),
  "userId-4" to listOf(" 03 060 701 "),
  "userId-5" to listOf(" 03-060-701"),
  "userId-7" to listOf(")'.03-060-701")
))
```

After the migration `newPhoneNumbers` will be equal to:

```kotlin
mapOf(
  "userId-1" to listOf("002250708060709"),
  "userId-2" to listOf("002250506060709"),
  "userId-6" to listOf("", "002250101020304"),
  "userId-3" to listOf("002250103060701"),
  "userId-4" to listOf("002250103060701"),
  "userId-5" to listOf("002250103060701"),
  "userId-7" to listOf("")
)
```

Invalid phone numbers are replaced by empty string.

## Download

Download the [latest JAR](https://search.maven.org/search?q=g:com.ouattararomuald%20AND%20a:numbering) or grab via Gradle:

```groovy
implementation 'com.ouattararomuald:numbering:2.0.1'
```

or Maven:

```xml
<dependency>
  <groupId>com.ouattararomuald</groupId>
  <artifactId>numbering</artifactId>
  <version>2.0.1</version>
</dependency>
```

Snapshots of the development version are available in [Sonatype's snapshots repository](https://oss.sonatype.org/content/repositories/snapshots/).

## Known limitations

- [x] Migrates only valid phone numbers.
- [x] Can add new digits before or after the old phone number. We decided to not handle in between insertions.
- [x] Can only look for digits to replace at the start, or the end of old phone numbers. We think this shouldn’t be a problem (generally) but as of now we decided to not handle such cases.
- [x] It is synchronous. This is a choice that will let you pick the any library you want to handle async tasks.

## Contributing

Contributions you say? Yes please!

**Bug report?**

If at all possible, please attach a minimal sample project or code which reproduces the bug.
Screenshots are also a huge help if the problem is visual.

**Send a pull request!**

If you're fixing a bug, please add a failing test or code that can reproduce the issue.

## License

```
Copyright 2020 Ouattara Gninlikpoho Romuald

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [2.0.1] - 2021-01-03
**Breaking change**:
- Replace invalid phone numbers and replace them by empty string. That change will allow
  users to know which phone number did not pass the migration.

## [1.0.1] - 2021-01-02
**Breaking change**:
- Migration of multiple phone numbers per user ID.

## [0.1]

Initial release.


src/test/kotlin/com/ouattararomuald/numbering/CountryPlanTest.kt
package com.ouattararomuald.numbering

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class CountryPlanTest {

  @BeforeEach
  fun setUp() {
  }

  @AfterEach
  fun tearDown() {
  }

  @Test
  fun `invalid old phone number size should fail`() {
    assertThrows<IllegalStateException>("Should throw IllegalStateException") {
      CountryPlan.Builder()
        .setOldPhoneNumberSize(1)
        .setInternationalCallingCode("225")
        .setMigrationType(MigrationType.POSTFIX)
        .setDigitMapperPosition(Position.START)
        .setPrefixesMapper(
          mapOf(
            "07" to "07",
            "08" to "07",
            "09" to "07",
            "04" to "05", // MTN
            "05" to "05",
            "06" to "05",
            "01" to "01", // MOOV
            "02" to "01",
            "03" to "01"
          )
        ).build()
    }
  }

  @Test
  fun `invalid international calling code should fail`() {
    assertThrows<IllegalStateException>("Should throw IllegalStateException") {
      CountryPlan.Builder()
        .setOldPhoneNumberSize(8)
        .setInternationalCallingCode(" ")
        .setMigrationType(MigrationType.POSTFIX)
        .setDigitMapperPosition(Position.START)
        .setPrefixesMapper(
          mapOf(
            "01" to "01",
            "02" to "01",
            "03" to "01"
          )
        ).build()
    }

    assertThrows<IllegalStateException>("Should throw IllegalStateException") {
      CountryPlan.Builder()
        .setOldPhoneNumberSize(8)
        .setInternationalCallingCode("zaeazeaze")
        .setMigrationType(MigrationType.POSTFIX)
        .setDigitMapperPosition(Position.START)
        .setPrefixesMapper(
          mapOf(
            "07" to "07",
            "08" to "07",
          )
        ).build()
    }

    assertThrows<IllegalStateException>("Should throw IllegalStateException") {
      CountryPlan.Builder()
        .setOldPhoneNumberSize(8)
        .setInternationalCallingCode("2 2 5")
        .setMigrationType(MigrationType.POSTFIX)
        .setDigitMapperPosition(Position.START)
        .setPrefixesMapper(
          mapOf(
            "07" to "07",
            "08" to "07",
          )
        ).build()
    }

    assertThrows<IllegalStateException>("Should throw IllegalStateException") {
      CountryPlan.Builder()
        .setOldPhoneNumberSize(8)
        .setInternationalCallingCode("225")
        .setMigrationType(MigrationType.POSTFIX)
        .setDigitMapperPosition(Position.START)
        .setPrefixesMapper(emptyMap()).build()
    }
  }
}

src/test/kotlin/com/ouattararomuald/numbering/NumberingPlanTest.kt
package com.ouattararomuald.numbering

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class NumberingPlanTest {

  private lateinit var numberingPlan: NumberingPlan

  @BeforeEach
  fun setUp() {
  }

  @AfterEach
  fun tearDown() {
  }

  @Test
  fun `PREFIX migration should keep only valid numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("08060709"),
        "userId-2" to PhoneNumber("06060709"),
        "userId-3" to PhoneNumber("03060701"),
        "userId-4" to PhoneNumber(" 03 060 701 "),
        "userId-5" to PhoneNumber(" 03-060-701"),
        "userId-6" to PhoneNumber("zezae/03-060-701"),
        "userId-7" to PhoneNumber(")'.03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250708060709"),
        "userId-2" to PhoneNumber("002250506060709"),
        "userId-3" to PhoneNumber("002250103060701"),
        "userId-4" to PhoneNumber("002250103060701"),
        "userId-5" to PhoneNumber("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration should keep only valid numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("08060709"),
        "userId-2" to PhoneNumber("06060709"),
        "userId-3" to PhoneNumber("03060701"),
        "userId-4" to PhoneNumber(" 03 060 701 "),
        "userId-5" to PhoneNumber(" 03-060-701"),
        "userId-6" to PhoneNumber("zezae/03-060-701"),
        "userId-7" to PhoneNumber(")'.03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250806070907"),
        "userId-2" to PhoneNumber("002250606070905"),
        "userId-3" to PhoneNumber("002250306070101"),
        "userId-4" to PhoneNumber("002250306070101"),
        "userId-5" to PhoneNumber("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration of phone numbers without icc`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)

    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("08060709"),
        "userId-2" to PhoneNumber("06060709"),
        "userId-3" to PhoneNumber("03060701"),
        "userId-4" to PhoneNumber(" 03 060 701 "),
        "userId-5" to PhoneNumber(" 03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250708060709"),
        "userId-2" to PhoneNumber("002250506060709"),
        "userId-3" to PhoneNumber("002250103060701"),
        "userId-4" to PhoneNumber("002250103060701"),
        "userId-5" to PhoneNumber("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration of phone numbers without icc`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)

    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("08060709"),
        "userId-2" to PhoneNumber("06060709"),
        "userId-3" to PhoneNumber("03060701"),
        "userId-4" to PhoneNumber(" 03 060 701 "),
        "userId-5" to PhoneNumber(" 03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250806070907"),
        "userId-2" to PhoneNumber("002250606070905"),
        "userId-3" to PhoneNumber("002250306070101"),
        "userId-4" to PhoneNumber("002250306070101"),
        "userId-5" to PhoneNumber("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration with valid phone numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("+22508060709"),
        "userId-2" to PhoneNumber("+22506060709"),
        "userId-3" to PhoneNumber("0022503060701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250708060709"),
        "userId-2" to PhoneNumber("002250506060709"),
        "userId-3" to PhoneNumber("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration with valid phone numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber("+22508060709"),
        "userId-2" to PhoneNumber("+22506060709"),
        "userId-3" to PhoneNumber("0022503060701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250806070907"),
        "userId-2" to PhoneNumber("002250606070905"),
        "userId-3" to PhoneNumber("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration of phone numbers with invalid chars`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber(" 00 22503 060 701 "),
        "userId-2" to PhoneNumber(" 00 225-03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250103060701"),
        "userId-2" to PhoneNumber("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration of phone numbers with invalid chars`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to PhoneNumber(" 00 22503 060 701 "),
        "userId-2" to PhoneNumber(" 00 225-03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to PhoneNumber("002250306070101"),
        "userId-2" to PhoneNumber("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  private fun generateCountryPlan(migrationType: MigrationType) = CountryPlan.Builder()
    .setOldPhoneNumberSize(8)
    .setInternationalCallingCode("225")
    .setMigrationType(migrationType) // MigrationType.prefix | MigrationType.postfix
    .setDigitMapperPosition(Position.START)
    .setPrefixesMapper(
      mapOf(
        "07" to "07", // Orange
        "08" to "07", // eg: 08 XX XX XX => 07 08 XX XX XX (if MigrationType.prefix is used) => 08 XX XX XX 07 (if MigrationType.postfix is used)
        "09" to "07",
        "04" to "05", // MTN
        "05" to "05",
        "06" to "05",
        "01" to "01", // MOOV
        "02" to "01",
        "03" to "01"
      )
    ).build()
}

numbering/src/test/kotlin/com/ouattararomuald/numbering/NumberingPlanTest.kt
package com.ouattararomuald.numbering

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class NumberingPlanTest {

  private lateinit var numberingPlan: NumberingPlan

  @BeforeEach
  fun setUp() {
  }

  @AfterEach
  fun tearDown() {
  }

  @Test
  fun `PREFIX migration should empty invalid numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("08060709"),
        "userId-2" to listOf("06060709"),
        "userId-6" to listOf("zezae/03-060-701", "01020304"),
        "userId-3" to listOf("03060701"),
        "userId-4" to listOf(" 03 060 701 "),
        "userId-5" to listOf(" 03-060-701"),
        "userId-7" to listOf(")'.03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250708060709"),
        "userId-2" to listOf("002250506060709"),
        "userId-6" to listOf("", "002250101020304"),
        "userId-3" to listOf("002250103060701"),
        "userId-4" to listOf("002250103060701"),
        "userId-5" to listOf("002250103060701"),
        "userId-7" to listOf("")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration should empty invalid numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("08060709"),
        "userId-2" to listOf("06060709"),
        "userId-3" to listOf("03060701"),
        "userId-4" to listOf(" 03 060 701 "),
        "userId-5" to listOf(" 03-060-701"),
        "userId-6" to listOf("zezae/03-060-701"),
        "userId-7" to listOf(")'.03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250806070907"),
        "userId-2" to listOf("002250606070905"),
        "userId-3" to listOf("002250306070101"),
        "userId-4" to listOf("002250306070101"),
        "userId-5" to listOf("002250306070101"),
        "userId-6" to listOf(""),
        "userId-7" to listOf("")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration of phone numbers without icc`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)

    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("08060709"),
        "userId-2" to listOf("06060709"),
        "userId-3" to listOf("03060701"),
        "userId-4" to listOf(" 03 060 701 "),
        "userId-5" to listOf(" 03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250708060709"),
        "userId-2" to listOf("002250506060709"),
        "userId-3" to listOf("002250103060701"),
        "userId-4" to listOf("002250103060701"),
        "userId-5" to listOf("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration of phone numbers without icc`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)

    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("08060709"),
        "userId-2" to listOf("06060709"),
        "userId-3" to listOf("03060701"),
        "userId-4" to listOf(" 03 060 701 "),
        "userId-5" to listOf(" 03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250806070907"),
        "userId-2" to listOf("002250606070905"),
        "userId-3" to listOf("002250306070101"),
        "userId-4" to listOf("002250306070101"),
        "userId-5" to listOf("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration with valid phone numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("+22508060709"),
        "userId-2" to listOf("+22506060709"),
        "userId-3" to listOf("0022503060701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250708060709"),
        "userId-2" to listOf("002250506060709"),
        "userId-3" to listOf("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration with valid phone numbers`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf("+22508060709"),
        "userId-2" to listOf("+22506060709"),
        "userId-3" to listOf("0022503060701"),
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250806070907"),
        "userId-2" to listOf("002250606070905"),
        "userId-3" to listOf("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `PREFIX migration of phone numbers with invalid chars`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.PREFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf(" 00 22503 060 701 "),
        "userId-2" to listOf(" 00 225-03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250103060701"),
        "userId-2" to listOf("002250103060701")
      ),
      formattedPhoneNumbers
    )
  }

  @Test
  fun `POSTFIX migration of phone numbers with invalid chars`() {
    val ivoryCoastPlanFactory: CountryPlan = generateCountryPlan(MigrationType.POSTFIX)
    numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
    val formattedPhoneNumbers = numberingPlan.migrate(
      mapOf(
        "userId-1" to listOf(" 00 22503 060 701 "),
        "userId-2" to listOf(" 00 225-03-060-701")
      )
    )

    assertEquals(
      mapOf(
        "userId-1" to listOf("002250306070101"),
        "userId-2" to listOf("002250306070101")
      ),
      formattedPhoneNumbers
    )
  }

  private fun generateCountryPlan(migrationType: MigrationType) = CountryPlan.Builder()
    .setOldPhoneNumberSize(8)
    .setInternationalCallingCode("225")
    .setMigrationType(migrationType) // MigrationType.prefix | MigrationType.postfix
    .setDigitMapperPosition(Position.START)
    .setPrefixesMapper(
      mapOf(
        "07" to "07", // Orange
        "08" to "07", // eg: 08 XX XX XX => 07 08 XX XX XX (if MigrationType.prefix is used) => 08 XX XX XX 07 (if MigrationType.postfix is used)
        "09" to "07",
        "04" to "05", // MTN
        "05" to "05",
        "06" to "05",
        "01" to "01", // MOOV
        "02" to "01",
        "03" to "01"
      )
    ).build()
}

numbering/src/main/kotlin/com/ouattararomuald/numbering/NumberingPlan.kt
package com.ouattararomuald.numbering

class NumberingPlan(val countryPlan: CountryPlan) {

  private val internationalCallingCodePattern =
    "^(\\+|0{2})?(?<icc>${countryPlan.internationalCallingCode})(?<phoneNumber>\\d{${countryPlan.oldPhoneNumberSize}})".toPattern()

  /**
   * Migrates the valid [phoneNumbers] to the new numbering plan.
   *
   * example:
   *
   * ```kotlin
   * val ivoryCoastPlanFactory: CountryPlan = CountryPlan.Builder()
   *   .setOldPhoneNumberSize(8)
   *   .setInternationalCallingCode("225")
   *   .setMigrationType(MigrationType.PREFIX)
   *   .setDigitMapperPosition(Position.START)
   *   .setPrefixesMapper(mapOf(
   *     "07" to "07",
   *     "08" to "07",
   *     "09" to "07",
   *     "04" to "05",
   *     "05" to "05",
   *     "06" to "05",
   *     "01" to "01",
   *     "02" to "01",
   *     "03" to "01"
   *   )).build()
   * val numberingPlan = NumberingPlan(ivoryCoastPlanFactory)
   * val formattedPhoneNumbers = numberingPlan.migrate(mapOf(
   *   "userId-1" to listOf("08060709"),
   *   "userId-2" to listOf("06060709"),
   *   "userId-3" to listOf("03060701"),
   *   "userId-4" to listOf(" 03 060 701 "),
   *   "userId-5" to listOf(" 03-060-701"),
   *   "userId-6" to listOf("zezae/03-060-701"),
   *   "userId-7" to listOf(")'.03-060-701")
   * ))
   * ```
   *
   * After the migration `formattedPhoneNumbers` will be equal to:
   * ```kotlin
   * mapOf(
   *   "userId-1" to listOf("002250806070907"),
   *   "userId-2" to listOf("002250606070905"),
   *   "userId-3" to listOf("002250306070101"),
   *   "userId-4" to listOf("002250306070101"),
   *   "userId-5" to listOf("002250306070101")
   * )
   * ```
   *
   * Invalid phone numbers will not pass the migration.
   *
   * @param phoneNumbers key-values list of phone numbers to migrate.
   * @param Key type of the key associated with each phone number.
   *
   * @throws MigrationException when fails to extract International Calling Code.
   * @throws MigrationException when fails to extract the Phone Number.
   *
   * @return the phone numbers whose migration was successful.
   */
  @Throws(MigrationException::class)
  fun <Key> migrate(phoneNumbers: Map<Key, List<String>>): Map<Key, List<String>> {
    val adaptedPhoneNumbers = phoneNumbers.map { it.key to it.value.map { number -> PhoneNumber(number) } }

    val outputPhoneNumbers: LinkedHashMap<Key, List<String>> = LinkedHashMap()
    adaptedPhoneNumbers.forEach { (key, givenPhoneNumbers) ->
      val cleanedPhoneNumbers = givenPhoneNumbers.map { givenPhoneNumber -> cleanPhoneNumber(givenPhoneNumber) }
      val formattedPhoneNumbers = mutableListOf<PhoneNumber>()
      cleanedPhoneNumbers.forEach { cleanedPhoneNumber ->
        if (cleanedPhoneNumber.isValidPhoneNumber() && cleanedPhoneNumber.isValidLength()) {
          formattedPhoneNumbers.add(formatPhoneNumber(cleanedPhoneNumber))
        } else {
          formattedPhoneNumbers.add(PhoneNumber(""))
        }
      }
      if (formattedPhoneNumbers.isNotEmpty()) {
        outputPhoneNumbers[key] = formattedPhoneNumbers.map { formattedPhoneNumber -> formattedPhoneNumber.value }
      }
    }
    return outputPhoneNumbers
  }

  private fun formatPhoneNumber(phoneNumber: PhoneNumber): PhoneNumber {
    return if (phoneNumber.startsWithInternationalCallingCode()) {
      val icc = phoneNumber.extractICC()
      val number = phoneNumber.extractPhoneNumberWithoutICC().addNewPrefix()
      PhoneNumber("00$icc$number")
    } else {
      val icc = countryPlan.internationalCallingCode
      PhoneNumber("00$icc${phoneNumber.value.addNewPrefix()}")
    }
  }

  private fun String.addNewPrefix(): String { // TODO: Refactor
    val keys = countryPlan.prefixesMapper.keys

    var hasMatched = false
    var phone = this

    for (key in keys) {
      val prefix = countryPlan.prefixesMapper[key]
      phone = if (canRunPrefixMigration(key)) {
        hasMatched = true
        migratePrefix(prefix, phone)
      } else {
        phone
      }
      if (hasMatched) {
        break
      }
    }

    return phone
  }

  private fun String.canRunPrefixMigration(key: String): Boolean {
    return (countryPlan.digitMapperPosition == Position.START && this.startsWith(key))
        || (countryPlan.digitMapperPosition == Position.END && this.endsWith(key))
  }

  private fun migratePrefix(prefix: String?, phone: String): String {
    if (prefix == null) {
      return phone
    }

    return if (countryPlan.migrationType == MigrationType.PREFIX) {
      "$prefix$phone"
    } else {
      "$phone$prefix"
    }
  }

  private fun PhoneNumber.isValidLength(): Boolean {
    return value.length >= countryPlan.oldPhoneNumberSize
  }

  private fun PhoneNumber.startsWithInternationalCallingCode(): Boolean {
    return internationalCallingCodePattern.matcher(value).matches()
  }

  /**
   * Extracts the International Calling Code from the phone number.
   *
   * @throws MigrationException when fails to extract International Calling Code.
   *
   * @return the international calling code.
   */
  @Throws(MigrationException::class)
  private fun PhoneNumber.extractICC(): String {
    val matcher = internationalCallingCodePattern.matcher(value)
    if (matcher.find()) {
      return matcher.group(ICC_GROUP_NAME)
    }
    throw MigrationException("Failed to extract International Calling Code") // Should never happen
  }

  /**
   * Extracts the phone number.
   *
   * If the phone number has international calling code (ICC), this will return the number without
   * the ICC.
   *
   * @throws MigrationException when fails to extract the Phone Number.
   *
   * @return the phone number.
   */
  @Throws(MigrationException::class)
  private fun PhoneNumber.extractPhoneNumberWithoutICC(): String {
    val matcher = internationalCallingCodePattern.matcher(value)
    if (matcher.find()) {
      return matcher.group(PHONE_NUMBER_GROUP_NAME)
    }
    throw MigrationException("Failed to extract Phone Number") // Should never happen
  }

  /**
   * Removes invalid characters from phone numbers.
   *
   * @param phoneNumber the phone number to clean.
   *
   * @return the cleaned phone number.
   */
  private fun cleanPhoneNumber(phoneNumber: PhoneNumber): PhoneNumber {
    return PhoneNumber(phoneNumber.value.replace(SPACE_REGEX, EMPTY_STRING))
  }

  private fun PhoneNumber.isValidPhoneNumber(): Boolean {
    return PHONE_PATTERN_1_REGEX.matches(value) || PHONE_PATTERN_2_REGEX.matches(value) || PHONE_PATTERN_3_REGEX.matches(
      value
    )
  }

  companion object {
    private const val EMPTY_STRING = ""

    private const val ICC_GROUP_NAME = "icc"
    private const val PHONE_NUMBER_GROUP_NAME = "phoneNumber"

    private const val PHONE_PATTERN_1 = "^(?<phoneNumber>\\d+)"
    private const val PHONE_PATTERN_2 = "^\\+(?<phoneNumber>\\d+)"
    private const val PHONE_PATTERN_3 = "^0{2}(?<phoneNumber>\\d+)"

    private val PHONE_PATTERN_1_REGEX = PHONE_PATTERN_1.toRegex()
    private val PHONE_PATTERN_2_REGEX = PHONE_PATTERN_2.toRegex()
    private val PHONE_PATTERN_3_REGEX = PHONE_PATTERN_3.toRegex()

    private const val PHONE_PATTERN = "^(\\+|00)\\d+"
    private val PHONE_REGEX = PHONE_PATTERN.toRegex()
    private val SPACE_REGEX = "[\\s\\-]+".toRegex()
  }
}
